File: leetcode/0053. Maximum Subarray/Solution.swift
// https://leetcode.com/problems/maximum-subarray/

class Solution {
    func maxSubArray(_ nums: [Int]) -> Int {
        var running_total = 0
        var result = Int.min
        for num in nums {
            if running_total < 0 {
                running_total = 0
            }
            running_total += num
            result = max(result, running_total)
        }

        return result
    }
}

let s = Solution()
assert(s.maxSubArray([-2,1,-3,4,-1,2,1,-5,4]) == 6)

File: leetcode/0053. Maximum Subarray/Solution.py
# https://leetcode.com/problems/maximum-subarray/

from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        if len(nums) == 1:
            return nums[0]

        current_maximum = nums[0]
        maximum = current_maximum
        
        for i in range(1, len(nums)):
            current_maximum = max(current_maximum + nums[i], nums[i])
            maximum = max(maximum, current_maximum)
        
        return maximum

s = Solution()
assert s.maxSubArray([-2,1,-3,4,-1,2,1,-5,4]) == 6
